extensions/eclipse-che-theia-workspace/src/browser/che-quick-open-workspace.ts

```diff
@@ -0,0 +1,229 @@
+import { inject, injectable } from 'inversify';
+import {
+    QuickOpenGroupItem,
+    QuickOpenItem,
+    QuickOpenMode,
+    QuickOpenModel
+} from '@theia/core/lib/common/quick-open-model';
+import { LabelProvider, QuickOpenService } from '@theia/core/lib/browser';
+import { CommandContribution, CommandRegistry } from '@theia/core/lib/common';
+import { WorkspaceCommands } from '@theia/workspace/lib/browser/workspace-commands';
+import { CheApiService } from '@eclipse-che/theia-plugin-ext/lib/common/che-protocol';
+import { che as cheApi } from '@eclipse-che/api';
+import { OauthUtils } from '@eclipse-che/theia-plugin-ext/lib/browser/oauth-utils';
+import { AbstractDialog } from '@theia/core/lib/browser/dialogs';
+import { MessageService } from '@theia/core/lib/common/message-service';
+import * as moment from 'moment';
+import { Message } from '@theia/core/lib/browser/widgets/index';
+import { Key } from '@theia/core/lib/browser/keyboard/keys';
+
+@injectable()
+export class QuickOpenCheWorkspace implements QuickOpenModel, CommandContribution {
+    protected items: QuickOpenGroupItem[];
+    protected currentWorkspace: cheApi.workspace.Workspace;
+
+    @inject(QuickOpenService) protected readonly quickOpenService: QuickOpenService;
+    @inject(CommandRegistry) protected readonly commandRegistry: CommandRegistry;
+    @inject(CheApiService) protected readonly cheApi: CheApiService;
+    @inject(OauthUtils) protected readonly oAuthUtils: OauthUtils;
+    @inject(LabelProvider) protected readonly labelProvider: LabelProvider;
+    @inject(MessageService) protected readonly messageService: MessageService;
+
+    registerCommands(commands: CommandRegistry): void {
+        // VS Code command id
+        commands.unregisterCommand('workbench.action.openRecent');
+
+        // Theia command id
+        commands.unregisterCommand('workspace:openRecent');
+
+        commands.registerCommand(WorkspaceCommands.OPEN_RECENT_WORKSPACE, {
+            execute: () => this.select()
+        });
+        commands.registerCommand({
+            id: 'workspace:openRecent'
+        }, {
+            execute: () => this.select()
+        });
+    }
+
+    protected async open(workspaces: cheApi.workspace.Workspace[]): Promise<void> {
+        this.items = [];
+
+        if (!workspaces.length) {
+            this.items.push(new QuickOpenGroupItem({
+                label: 'No Recent Workspaces',
+                run: (mode: QuickOpenMode): boolean => false
+            }));
+            return;
+        }
+
+        for (const workspace of workspaces) {
+            const icon = this.labelProvider.folderIcon;
+            const iconClass = icon + ' file-icon';
+            this.items.push(new QuickOpenGroupItem({
+                label: this.getWorkspaceName(workspace) + (this.isCurrentWorkspace(workspace) ? ' (Current)' : ''),
+                description: this.getWorkspaceStack(workspace),
```

Comment:
Since a Stack can be a long string, in case of Devfile link for example,
maybe it worth displaying it as [`detail`](https://github.com/eclipse-theia/theia/blob/a1ea2307d50565944b42f74ba3797698697ec897/packages/core/src/common/quick-open-model.ts#L38) instead of `description`?

It should allow using the whole widget width for Stack name, e.g.:
![image](https://user-images.githubusercontent.com/1636395/87396807-10869380-c5bc-11ea-94c0-db7fe0f0e19c.png)

Reply:
Done.